pageserver/src/tenant/timeline.rs

```diff
@@ -6522,10 +6554,29 @@ impl Timeline {
         };
 
         let mut new_gc_cutoff = space_cutoff.min(time_cutoff.unwrap_or_default());
-        let standby_horizon = self.standby_horizon.load();
+
         // Hold GC for the standby, but as a safety guard do it only within some
         // reasonable lag.
-        if standby_horizon != Lsn::INVALID {
+        // TODO: revisit this once we've fully transitioned to leases. 10GiB isn't _that_ much
+        // of permitted lag at high ingest rates. Yet again, unlimited lag is a DoS risk.
+        // Ideally we implement a monitoring function for RO replica lag in a higher level controller,
+        // e.g. the TBD compute manager aka endpoint controller. It would kill the replica if it's
+        // lagging too much or is otherwise unresponsive. The standby horizon leasing could
+        // then be moved into that controller, i.e., inside our trust domain. Replica can still
+        // advance existing leases, but, control over existence (create/destroy lease) would be
+        // with that controller.
+        // When solving this, solve it generically for lsn leases as well.
+        let min_standby_horizon = self.standby_horizons.min_and_clear_legacy();
+        let min_standby_horizon = if self
+            .feature_resolver
+            .evaluate_boolean("standby-horizon-leases-in-gc")
+            .is_ok()
+        {
+            min_standby_horizon.all
+        } else {
+            min_standby_horizon.legacy
+        };
+        if let Some(standby_horizon) = min_standby_horizon {
             if let Some(standby_lag) = new_gc_cutoff.checked_sub(standby_horizon) {
```

Comment:
NB pre-exisiting bug I spotted here: this subtraction can succeed with Some() and if it's less than the max lag, we'll set `new_gc_cutoff` to the `standby_horizon`. However, nothing prevents that value from going below applied gc cutoff, which will then make the `self.gc_timeline()` call below bail with an error.

I guess we've been lucky so far?

Reply:
I added 3-way feature gating (legacy, all, leases-only) in this commit https://github.com/neondatabase/neon/pull/12676/commits/73336962a8bb6efc4e05b124d0bab8b2db425016

Once we are leases-only, this bug can't happen anymore. Until then, the feature gate evaluation is logged, so if this rare bug should occur, we can tell from logs whether `legacy` was involved or not.